### DISCUSSION TITLE
What memory allocation does vips use

### DISCUSSION THREAD

**Discussion by zhaohuxing** [libvips/libvips#2917] [Help wanted]
Hi @jcupitt  longlongtime, I use rust binding vips, my program keeps crashing, but I can't find the problem, what memory allocation does vips use? In addition, what should be paid attention to when using vips, such as what configuration to open and so on

**Comment by jcupitt** [comment]
Hi @zhaohuxing,

libvips uses the standard system malloc.

You would need to post some code, but crashes are usually if you free to early, or unref too many times. 

I would read the documentation, and perhaps look at the Python or Ruby or C++ bindings for examples.

**Comment by zhaohuxing** [comment]
Hi @jcupitt, I run `cargo test` in [vips-image](https://github.com/zhaohuxing/vips-image), print error log
```
unning 2 tests

(process:14779): GLib-GObject-WARNING **: 19:07:30.550: cannot retrieve class for invalid (unclassed) type '<invalid>'

(process:14779): GLib-GObject-WARNING **: 19:07:30.551: invalid class cast from (NULL) pointer to 'VipsObject'

(process:14779): GLib-GObject-WARNING **: 19:07:30.551: invalid class cast from (NULL) pointer to 'VipsObject'
error: test failed, to rerun pass '--test watermark'

Caused by:
  process didn't exit successfully: `/Users/xzh/workspaces/rust/playground-rust/vips-image/target/debug/deps/watermark-4632591f8e179820` (signal: 11, SIGSEGV: invalid memory reference)
```

**Reply by jcupitt** [reply]
libvips has two C APIs and it looks like you're using the human-oriented one.

libvips has a second C API designed for language bindings. This is fully typesafe (no varargs!) and supports introspection, so you can generate most of the API at runtime. I would use this binding API instead. It should be much easier.

There's an example here:

https://www.libvips.org/API/current/binding.html

**Reply by jcupitt** [reply]
Looking at the error messages, perhaps you are unreffing too early, or you have not called `vips_init()` to start the library up.

**Reply by zhaohuxing** [reply]
I use [libvips-rust-bindings](https://github.com/olxgroup-oss/libvips-rust-bindings), Do you mean it should call the lower level c API?

I call vips_init, and also print same error.

**Reply by jcupitt** [reply]
Ah OK, I thought you were calling libvips directly.

I don't know anything about libvips-rust-bindings, but I suppose it's all OK, it seems like a mature project.

I would ask about your crashes there.
